pennylane/transforms/add_noise.py

```diff
@@ -0,0 +1,320 @@
+        TransformProgram(add_noise)
+        >>> qml.transforms.add_noise(circuit, noise_model, level="user").transform_program
+        TransformProgram(cancel_inverses, merge_rotations, undo_swaps, _expand_metric_tensor, add_noise)
+        >>> qml.transforms.add_noise(circuit, noise_model, level="device").transform_program
+        TransformProgram(cancel_inverses, merge_rotations, undo_swaps, _expand_metric_tensor, batch_transform, expand_fn, add_noise)
+
+        Finally, more precise control over the insertion of the transform can be achieved by specifying
+        an integer or slice for indexing for extracting the transform program. For example, one can do:
+
+        >>> qml.transforms.add_noise(circuit, noise_model, level=2).transform_program
+        TransformProgram(cancel_inverses, merge_rotations, add_noise)
+        >>> qml.transforms.add_noise(circuit, noise_model, level=slice(1,3)).transform_program
+        TransformProgram(merge_rotations, undo_swaps, add_noise)
+
+    """
+    if not hasattr(noise_model, "model_map") or not hasattr(noise_model, "metadata"):
+        raise ValueError(
+            f"Provided noise model object must define model_map and metatadata attributes, got {noise_model}."
+        )
+
+    if level is None or level == "user":  # decompose templates and their adjoints
+
+        def stop_at(obj):
+            if not isinstance(obj, qml.operation.Operator):
+                return True
+            if not obj.has_decomposition:
+                return True
+            return not (hasattr(qml.templates, obj.name) or isinstance(obj, qml.ops.Adjoint))
+
+        error_type = (qml.operation.DecompositionUndefinedError,)
+        decompose = qml.devices.preprocess.decompose
+        [tape], _ = decompose(tape, stopping_condition=stop_at, name="add_noise", error=error_type)
+
+    conditions, noises = [], []
+    metadata = noise_model.metadata
+    for condition, noise in noise_model.model_map.items():
+        conditions.append(lru_cache(maxsize=512)(condition))
+        noises.append(qml.tape.make_qscript(noise))
+
+    new_operations = []
+    for operation in tape.operations:
+        curr_ops = [operation]
+        for condition, noise in zip(conditions, noises):
+            if condition(operation):
+                noise_ops = noise(operation, **metadata).operations
+                if operation in noise_ops and _check_queue_op(operation, noise, metadata):
+                    ops_indx = noise_ops.index(operation)
+                    curr_ops = noise_ops[:ops_indx] + curr_ops + noise_ops[ops_indx + 1 :]
+                else:
+                    curr_ops.extend(noise_ops)
+        new_operations.extend(curr_ops)
+
+    new_tape = type(tape)(new_operations, tape.measurements, shots=tape.shots)
+    post_processing_fn = qml.devices.preprocess.null_postprocessing
+
+    return [new_tape], post_processing_fn
+
+
+def _check_queue_op(operation, noise_func, metadata):
```

Comment:
This function is needed to allow flexibility in queuing the noise operations. This allows checking whether an operation exists in the `queue` obtained from a `noise_fn`. If so, we remove it and simply follow the gate sequence defined by the `noise_fn`. 

For doing such a check, we cannot simply rely on something like `qml.equal` since it doesn't distinguish between two operations based on their `id`. Instead, it just checks their parameters, which means if the `noise_fn` somehow produced the same operation as that of the `op` being evaluated by the condition, it will result in a False positive.